### ISSUE TITLE
Bpatch  attach process takes long time and requires more meory

### ISSUE THREAD

**Issue by ManojKumarChauhan** [dyninst/dyninst#1188]
processAttach function of BPatch class take 5-10 minutes to attach to a big size  binary and requires 5-10 gb of free memory.

bpatch.processAttach (NULL, pid);
here pid is the process id of any c/c++/golang process. 

processAttach function takes min 2 minute for a small process and requires 2 gb of memory. 
As the process size increased, time and memory required for processAttach also increased.  For large process it takes 5-10 minutes and requires 5-10 gb of free RAM.
Is there any option to reduce time and memory?

**Comment by hainest** [comment]
Can you provide the following?

1. the system and Dyninst information requested in the issue template
2. a minimum working example of the mutator used to attach to the process
3. the type of process being attached to (user, system, daemon/zombie, etc.)
4. the output of running your mutator with `DYNINST_DEBUG_DYNPC=1` set in the environment

**Comment by ManojKumarChauhan** [comment]
1. OS
NAME="Ubuntu"
VERSION="16.04.2 LTS (Xenial Xerus)"
Dynist version 10.2.1

2. Attached code of mutator
[codeCoverage.cpp.txt](https://github.com/dyninst/dyninst/files/7925316/codeCoverage.cpp.txt)
run it using ./codeCoverage inputbinaryname
I used it on some large c++ application. Also tried on a go application influxd (attached)
3. process type user
4. ./codeCoverage influxd
Enabling DyninstAPI process control debug
pid of process 27125
24-01-2022 17:28:10[140072784304576]dynProcess.C[1483]: added thread 27125 to process 27125
[140072784304576]dynProcess.C[1483]: added thread 27126 to process 27125
[140072784304576]dynProcess.C[1483]: added thread 27127 to process 27125
[140072784304576]dynProcess.C[1483]: added thread 27128 to process 27125
[140072784304576]dynProcess.C[1483]: added thread 27129 to process 27125
[140072784304576]dynProcess.C[1483]: added thread 27130 to process 27125
[140072784304576]dynProcess.C[1483]: added thread 27137 to process 27125
[140072784304576]pcEventMuxer.C[232]: Begin callbackMux, process pointer = 0x2d7a1b60, event Library
[140072784304576]pcEventMuxer.C[494]: Added event Library from process 27125 to mailbox, size now 1
[140072784304576]--------- Enqueue for Process ID [27125] -------------
[140072784304576]	27125 -> 1
[140072784304576]---------------------------------------------------
[140072784304576]pcEventMuxer.C[234]: after muxing event, mailbox size is 1
[140072784304576]pcEventMuxer.C[552]: Returning event Library from mailbox for process 27125
[140072784304576]--------- Dequeue for Process ID [27125] -------------
[140072784304576]	27125 -> 0
[140072784304576]---------------------------------------------------
[140072784304576]pcEventHandler.C[67]: attempting to handle event Library on thread 27125/27125
[140072784304576]pcEventHandler.C[877]: new mapped object: libdyninstAPI_RT.so.10.2.1:371089408/31028
[140072784304576]pcEventHandler.C[888]: library event contains RT library load
[140072784304576]Calling find for process 0x2d7a1b60 (27125)
[140072784304576]pcEventHandler.C[192]: continue condition ( 1 0 1 1 1 1 )
[140072784304576]Calling find for process 0x2d7a1b60 (27125)
[codeCoverage.cpp.txt](https://github.com/dyninst/dyninst/files/7925397/codeCoverage.cpp.txt)

[140072784304576]pcEventHandler.C[228]: delaying process continue: <return success> <desired stop> <stopped> <no reported event> <not terminated> <no pending events>
[140072784304576]pcEventHandler.C[240]: finished handling event: Library (err = false)
attached to process influxd
[140072784304576]dynProcess.C[992]: Continuing process 27125

**Comment by ManojKumarChauhan** [comment]
I created tar for influxdb & the size of tar  is more than 100mb. Unable to attach.

**Comment by bigtrak** [comment]
Please share the following info so we can look into this further;
I'm taking a look at it, and the detail information you can provide 
can help us replicate your environment and identify the cause of the behavior.

First, ubuntu 16 is an older platform, which is not a problem.
Please indicate what process(es) are you using to install dyninst 
and it's dependencies, and also tell us which compiler are you using?

In particular, what are the versions of
   boost
   elfutils
   intel-tbb
   gcc
that you are using?


I'd also like to get an idea of your system's configuration, in case that affects the issue.
1) What  CPUs (processor) are you using?  How many CPUs and Cores?
2) How much memory does your system have?
3) According to top (or other  tools) how much free memory does your system have before
you start trying to run codeCoverage, and what happens to that free memory as codeCoverage
runs -- does the system start swapping, or does it just require a large amount of memory?

4) Which influxdb executable are you instrumenting -- one from a ubuntu repository or package (such
as by adding influx as a package source, or from ubuntu), compiled yourself
from github, or some other method?   If  you configured and compiled it, please
share the configuration options you used, and the compiler name and version
it was compiled with.   

These details will help us take a look at this issue to try and replicate it.

**Comment by ManojKumarChauhan** [comment]
I installed Dynist from source code. System where dynist is installed->gcc --version
gcc (Ubuntu 6.5.0-2ubuntu1~16.04) 6.5.0 20181026
boost-libboost_system.so.1.61.0
elf-libelf-0.178.so
tbb-libtbb.so.2

influxdb is also compiled from source code on same system using https://github.com/influxdata/influxdb/blob/master/CONTRIBUTING.md#building-from-source
There is no specific configuration for influxd. To login, I created user/pass and one token.

After installing dynist  and influxd, I copied all libs & bin on another system and did instrumentation there.  gcc version->
gcc --version
gcc (Ubuntu 5.4.0-6ubuntu1~16.04.12) 5.4.0 20160609

lscpu
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                4
On-line CPU(s) list:   0-3
Thread(s) per core:    2
Core(s) per socket:    2
Socket(s):             1
NUMA node(s):          1
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 142
Model name:            Intel(R) Core(TM) i3-8145U CPU @ 2.10GHz
Stepping:              12
CPU MHz:               703.471
CPU max MHz:           3900.0000
CPU min MHz:           400.0000
BogoMIPS:              4599.93
Virtualization:        VT-x
L1d cache:             32K
L1i cache:             32K
L2 cache:              256K
L3 cache:              4096K
NUMA node0 CPU(s):     0-3

memory before running codecoverage
free -h
              total        used        free      shared  buff/cache   available
Mem:           7.8G        1.8G        5.2G        110M        807M        5.6G
Swap:          974M        952M         22M


After running influxd
free -h
              total        used        free      shared  buff/cache   available
Mem:           7.8G        1.9G        4.9G        110M        926M        5.5G
Swap:          974M        952M         22M


memory used by dyninst'd process attach

free -h
              total        used        free      shared  buff/cache   available
Mem:           7.8G        5.4G        1.4G        110M        1.0G        2.0G
Swap:          974M        952M         22M


Let me know if you need any other details.
Total time used by processattach is approx 5 mins.

**Comment by ManojKumarChauhan** [comment]
If i run influxdb and dynist on same machine where I installed them then processattach takes 2 mins.
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                4
On-line CPU(s) list:   0-3
Thread(s) per core:    2
Core(s) per socket:    2
Socket(s):             1
NUMA node(s):          1
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 142
Model name:            Intel(R) Core(TM) i3-8145U CPU @ 2.10GHz
Stepping:              12
CPU MHz:               703.471
CPU max MHz:           3900.0000
CPU min MHz:           400.0000
BogoMIPS:              4599.93
Virtualization:        VT-x
L1d cache:             32K
L1i cache:             32K
L2 cache:              256K
L3 cache:              4096K
NUMA node0 CPU(s):     0-3

**Comment by ManojKumarChauhan** [comment]
> 
for machine (CPU(s): 4), process attach takes times between 1 min 45 seconds to 3 min. Sometimes it takes 1 min 45 sec and sometimes 3 min 2 sec. This is strange and not stable while the free memory is almost same in both the cases.

**Comment by bigtrak** [comment]
Thank you for the detailed information.   I am looking into your issue.  Running your mutator against the influxd   from the influxdb apt repository (version 18  IIRC) it works correctly.   Building influxdb latest release by hand has been problematic.   Do you have  a script or short series of commands that you use to build
your version on influxdb so that I can get closer to the issues you may be experiencing?

**Comment by ManojKumarChauhan** [comment]
I used latest code of influxdb and compiled it in go17. Followed this page to install
https://github.com/influxdata/influxdb/blob/master/CONTRIBUTING.md#building-from-source
1. install gvm
bash < <(curl -s -S -L https://raw.githubusercontent.com/moovweb/gvm/master/binscripts/gvm-installer)
gvm install go1.17
gvm use go1.17 --default
Set GO111MODULE=on in bashrc
2. sudo apt install make clang pkg-config protobuf-compiler libprotobuf-dev
3. git clone influxd
https://github.com/influxdata/influxdb.git
4. make
could you please share the link for version 18 IIRC

**Comment by ManojKumarChauhan** [comment]
hi @bigtrak ,
did you get chance to look at it?

**Comment by bigtrak** [comment]
Hello, I'm looking at some interesting side effects.  As far as using the distributed version of influxdb, just look for installing influx db from the influxdb apt repos.

However, please try something out; this may substantially reduce the memory use of running dyninst again
influxdb

export OMP_NUM_THREADS=1
then run your mutator against the influx DB as you tried before.

If you see something unexpected occur, such as a seg fault in the mutator, just a quick
reminder that the environment variable DYNINSTAPI_RT_LIB must be set correctly
for the mutator, and also that dyninst libs and the depencencies are in LD_LIBRARY_PATH.

I mention that, because it is quite easy to forget those, or run in a different window where they
are not set.

**Comment by ManojKumarChauhan** [comment]
DYNINSTAPI_RT_LIB and LD_LIBRARY_PATH are set correctly. Sometimes dynist is taking 8-10  mins on system with 4 cpu core and 16gb RAM.  I will try with OMP_NUM_THREADS

**Comment by ManojKumarChauhan** [comment]
after setting env variable OMP_NUM_THREADS, there was no seg fault and mutator took 5-6 mins in bpatch.processAttach

**Comment by ManojKumarChauhan** [comment]
Hi @bigtrak ,
There is some improvement in the latest code. I have merged 12.1.0. For system having 1 cpu core, attachprocess is taking 6-7 minutes while before it was taking 18-20 minutes. For system having 4+ cpu cores attach process is taking 2 minutes in latest code.

So this issue has been solved but I got a new issue and need your suggestion.

I wrapped some functions. After some time the application is crashing (stack got corrupt) when cpu core is 4+.
But the application does not crash when cpu core is 1.
Any suggestion where should I look?
